crates/uv/src/commands/pip/tree.rs

```diff
@@ -0,0 +1,210 @@
+// all times, `PySocks` on `socks` extra and `chardet` on `use_chardet_on_py3` extra.
+// This function will return `["charset-normalizer", "idna", "urllib", "certifi"]` for `requests`.
+fn required_with_no_extra(dist: &InstalledDist) -> Vec<pep508_rs::Requirement<VerbatimParsedUrl>> {
+    let metadata = dist.metadata().unwrap();
+    return metadata
+        .requires_dist
+        .into_iter()
+        .filter(|r| {
+            r.marker.is_none()
+                || !r
+                    .marker
+                    .as_ref()
+                    .unwrap()
+                    .evaluate_optional_environment(None, &metadata.provides_extras[..])
+        })
+        .collect::<Vec<_>>();
+}
+
+// Render the line for the given installed distribution in the dependency tree.
+fn render_line(installed_dist: &InstalledDist, indent: usize, is_visited: bool) -> String {
+    let mut line = String::new();
+    if indent > 0 {
+        line.push_str("    ".repeat(indent - 1).as_str());
+        line.push_str("└──");
+        line.push(' ');
+    }
+    line.push_str((*installed_dist.name()).to_string().as_str());
+    line.push_str(" v");
+    line.push_str((*installed_dist.version()).to_string().as_str());
+
+    if is_visited {
+        line.push_str(" (*)");
+    }
+    line
+}
+#[derive(Debug)]
+struct DisplayDependencyGraph<'a> {
+    site_packages: &'a SitePackages,
+    // Map from package name to the installed distribution.
+    dist_by_package_name: HashMap<&'a PackageName, &'a InstalledDist>,
+    // Set of package names that are required by at least one installed distribution.
+    // It is used to determine the starting nodes when recursing the
+    // dependency graph.
+    required_packages: HashSet<PackageName>,
+}
+
+impl<'a> DisplayDependencyGraph<'a> {
+    /// Create a new [`DisplayDependencyGraph`] for the set of installed distributions.
+    fn new(site_packages: &'a SitePackages) -> DisplayDependencyGraph<'a> {
+        let mut dist_by_package_name = HashMap::new();
+        let mut required_packages = HashSet::new();
+        for site_package in site_packages.iter() {
+            dist_by_package_name.insert(site_package.name(), site_package);
+        }
+        for site_package in site_packages.iter() {
+            for required in required_with_no_extra(site_package) {
+                required_packages.insert(required.name.clone());
+            }
+        }
+
+        Self {
+            site_packages,
+            dist_by_package_name,
+            required_packages,
+        }
+    }
+
+    // Depth-first traversal of the given distribution and its dependencies.
+    fn visit(
+        &self,
+        installed_dist: &InstalledDist,
+        visited: &mut HashSet<String>,
+        path: &mut Vec<String>,
+    ) -> Result<Vec<String>, Error> {
+        let mut lines = Vec::new();
+
+        let package_name = installed_dist.name().to_string();
```

Comment:
Same as above.

```suggestion
        let package_name = installed_dist.name().as_ref();
```

Reply:
Hmm looks like this was marked as resolved but didn't get changed?

Reply:
oh oops, yah i think i reverted this while working on other comments; this change makes compiler to complain about a flurry of other things and my tiny brain was not strong enough to win that battle - will look again after fixing the display issue 😅

Reply:
Ah you end up needing the `String` later anyways. You could use a `HashSet<&str>` and `Vec<&str>` for the `visited` and `lines` types to avoid the allocations, but it's fine how it is now.

Reply:
got it, yah maybe better to leave it as is for now then (was iffy around adding all the lifetimes annotation etc 😅 )